tachyon/math/finite_fields/small_prime_field_generic.h

```diff
@@ -0,0 +1,230 @@
+
+  constexpr bool IsZero() const { return value_ == 0; }
+
+  constexpr bool IsOne() const { return value_ == 1; }
+
+  std::string ToString() const { return base::NumberToString(value_); }
+
+  std::string ToHexString(bool pad_zero = false) const {
+    std::string str = base::HexToString(value_);
+    if (pad_zero) {
+      str = base::ToHexStringWithLeadingZero(str, 8);
+    }
+    return base::MaybePrepend0x(str);
+  }
+
+  mpz_class ToMpzClass() const {
+    mpz_class ret;
+    gmp::WriteLimbs(ToBigInt().limbs, N, &ret);
+    return ret;
+  }
+
+  // TODO(chokobole): Support bigendian.
+  constexpr BigInt<N> ToBigInt() const { return BigInt<N>(value_); }
+
+  constexpr MontgomeryTy ToMontgomery() const {
+    return MontgomeryTy(
+        Config::Reduce(uint64_t{value_} * Config::kMontgomeryR[0]));
+  }
+
+  constexpr operator uint32_t() const { return value_; }
+
+  uint32_t operator[](size_t i) const {
+    DCHECK_EQ(i, size_t{0});
+    return value_;
+  }
+
+  bool operator==(PrimeField other) const { return value_ == other.value_; }
+  bool operator!=(PrimeField other) const { return value_ != other.value_; }
+  bool operator<(PrimeField other) const { return value_ < other.value_; }
+  bool operator>(PrimeField other) const { return value_ > other.value_; }
+  bool operator<=(PrimeField other) const { return value_ <= other.value_; }
+  bool operator>=(PrimeField other) const { return value_ >= other.value_; }
+
+  // AdditiveSemigroup methods
+  constexpr PrimeField Add(PrimeField other) const {
+    return PrimeField(Config::Reduce(uint64_t{value_} + other.value_));
+  }
+
+  constexpr PrimeField& AddInPlace(PrimeField other) {
+    value_ = Config::Reduce(uint64_t{value_} + other.value_);
+    return *this;
+  }
+
+  // AdditiveGroup methods
+  constexpr PrimeField Sub(PrimeField other) const {
+    return PrimeField(
+        Config::Reduce(uint64_t{value_} + GetModulus() - other.value_));
+  }
+
+  constexpr PrimeField& SubInPlace(PrimeField other) {
+    value_ = Config::Reduce(uint64_t{value_} + GetModulus() - other.value_);
+    return *this;
+  }
+
+  constexpr PrimeField Negate() const {
+    return PrimeField(Config::Reduce(uint64_t{GetModulus()} - value_));
+  }
+
+  constexpr PrimeField& NegateInPlace() {
+    value_ = Config::Reduce(uint64_t{GetModulus()} - value_);
+    return *this;
+  }
+
+  // MultiplicativeSemigroup methods
+  constexpr PrimeField Mul(PrimeField other) const {
+// NOTE(chokobole): g++ 11.4.0 gives a warning. It seems a g++ bug.
+#if defined(COMPILER_GCC) && !defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif
+    return PrimeField(Config::Reduce(uint64_t{value_} * other.value_));
+#if defined(COMPILER_GCC) && !defined(__clang__)
+#pragma GCC diagnostic pop
+#endif
+  }
+
+  constexpr PrimeField& MulInPlace(PrimeField other) {
+    value_ = Config::Reduce(uint64_t{value_} * other.value_);
```

Comment:
Just curious, but what this`Reduce` does/how it will be implemented -> will it ensure that the final number is within the range of the modulus/ does it do more than that?

Reply:
Custom reduction method different from montgomery reduction? Basically, as you said, this ensures the number is within the range, but this is implemented in a fast way. Because every prime field under 64 bit can be generalized as belows, but my assumption is that there is a better reduction and this is why the field is chosen. for example, if the field is 2^31, the reduction will be just a bit operation: `value_ & (2^31 - 1)`

```c++
if (value_ >= Config::kModulus[0]) {
  value_ -= Config::kModulus[0];
}
```